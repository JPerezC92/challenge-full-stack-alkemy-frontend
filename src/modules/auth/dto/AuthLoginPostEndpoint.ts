import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";
import { UserEndpoint } from "./UserEndpoint";

export type AuthLoginPostEndpoint =
  | JsendSuccess<{
      accessToken: string;
      refreshToken: string;
      user: UserEndpoint;
    }>
  | JsendError
  | JsendFail;
