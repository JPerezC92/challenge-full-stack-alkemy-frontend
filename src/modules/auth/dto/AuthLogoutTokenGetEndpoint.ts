import {
  JsendError,
  JsendFail,
  JsendSuccess,
} from "src/modules/shared/service/JsendResponse";

export type AuthLogoutTokenGetEndpoint = JsendSuccess | JsendError | JsendFail;
