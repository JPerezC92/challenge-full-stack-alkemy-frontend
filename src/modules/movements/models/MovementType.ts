export enum MovementType {
  INCOME = "income",
  EXPENSE = "expense",
}
