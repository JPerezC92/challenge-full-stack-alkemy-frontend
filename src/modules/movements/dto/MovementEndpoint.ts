import { MovementType } from "src/modules/movements/models/MovementType";

export interface MovementEndpoint {
  id: string;
  concept: string;
  amount: number;
  date: Date;
  type: MovementType;
}
