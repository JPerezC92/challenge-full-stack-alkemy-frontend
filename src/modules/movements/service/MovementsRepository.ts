import { MovementCreateDto } from "src/modules/movements/dto/MovementCreateDto";
import { Movement } from "src/modules/movements/models/Movement";
import { MovementType } from "src/modules/movements/models/MovementType";
import { OrderType } from "src/modules/shared/models/OrderType";

export interface MovementsRepository {
  query(query: {
    limit?: number;
    movementType?: MovementType;
    order?: OrderType;
    page?: number;
  }): Promise<Movement[]>;
  create(movementCreateDto: Omit<Movement, "id">): Promise<void>;
}
