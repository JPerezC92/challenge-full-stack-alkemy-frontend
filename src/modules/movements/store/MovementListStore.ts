import { MovementEndpoint } from "../dto/MovementEndpoint";

export interface MovementListStore {
  updateMovementList: (movements: MovementEndpoint[]) => void;
}
