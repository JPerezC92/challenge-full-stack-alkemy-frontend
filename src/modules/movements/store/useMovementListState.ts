import React from "react";
import { MovementEndpoint } from "src/modules/movements/dto/MovementEndpoint";
import { MyStore } from "src/modules/shared/store/MyStore";
import { MovementListStore } from "./MovementListStore";

export function useMovementListState() {
  const [movementList, setMovementList] = React.useState<MovementEndpoint[]>(
    []
  );

  const movementListStore: MyStore<MovementListStore> = React.useCallback(
    () => ({
      updateMovementList: (movements: MovementEndpoint[]) =>
        setMovementList(movements),
    }),
    []
  );

  return { movementList, movementListStore };
}
