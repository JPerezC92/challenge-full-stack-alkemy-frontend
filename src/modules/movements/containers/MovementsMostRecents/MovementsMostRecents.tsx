import React from "react";
import { MovementRecentCard } from "src/modules/movements/components/MovementRecentCard";
import { MovementsRepository } from "src/modules/movements/service/MovementsRepository";
import { useMovementListState } from "src/modules/movements/store/useMovementListState";
import { useRequest } from "src/modules/shared/hooks/useRequest";
import { OrderType } from "src/modules/shared/models/OrderType";
import { MyRepository } from "src/modules/shared/service/MyRepository";

type MovementsMostRecentsProps = {
  movementsRepository: MyRepository<MovementsRepository>;
};

export const MovementsMostRecents: React.FC<MovementsMostRecentsProps> = ({
  movementsRepository,
}) => {
  const { movementList, movementListStore } = useMovementListState();

  useRequest(
    async ({ abortController }) => {
      const _movementsRepository = movementsRepository({ abortController });
      const _movementsListStore = movementListStore();

      const movementList = await _movementsRepository.query({
        limit: 10,
        order: OrderType.DESC,
      });

      if (!movementList) return;

      _movementsListStore.updateMovementList(movementList);
    },
    [movementListStore, movementsRepository]
  );

  return (
    <div>
      <h2>Recent movements</h2>

      <ol>
        {movementList.map((movement) => (
          <MovementRecentCard key={movement.id} {...movement} />
        ))}
      </ol>
    </div>
  );
};
