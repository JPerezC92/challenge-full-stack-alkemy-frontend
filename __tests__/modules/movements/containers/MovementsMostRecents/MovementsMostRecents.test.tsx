import { render, waitFor } from "@testing-library/react";
import renderer from "react-test-renderer";
import { MovementsMostRecents } from "src/modules/movements/containers/MovementsMostRecents";
import { MovementEndpoint } from "src/modules/movements/dto/MovementEndpoint";
import { MovementType } from "src/modules/movements/models/MovementType";
import { MovementsRepository } from "src/modules/movements/service/MovementsRepository";
import * as useNodeMovementsRepository from "src/modules/movements/service/useNodeMovements.repository";
import { MovementListStore } from "src/modules/movements/store/MovementListStore";
import * as useMovementListState from "src/modules/movements/store/useMovementListState";
import { OrderType } from "src/modules/shared/models/OrderType";

const movementList = Array(10)
  .fill("")
  .map(
    (_, index): MovementEndpoint => ({
      id: (index + 1).toString(),
      amount: 1000 * (index + 1),
      concept: "move",
      date: new Date(),
      type: MovementType.EXPENSE,
    })
  );

const movementListStore: MovementListStore = {
  updateMovementList: jest.fn(),
};

const movementsRepository: MovementsRepository = {
  create: jest.fn(),
  query: jest.fn().mockResolvedValue(movementList),
};

jest.spyOn(useMovementListState, "useMovementListState").mockReturnValue({
  movementList,
  movementListStore: () => movementListStore,
});
jest
  .spyOn(useNodeMovementsRepository, "useNodeMovementsRepository")
  .mockReturnValue(jest.fn().mockReturnValue(movementsRepository));

describe("MovementsMostRecents container", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test("should match the snapshot", () => {
    const component = renderer
      .create(
        <MovementsMostRecents
          movementsRepository={useNodeMovementsRepository.useNodeMovementsRepository()}
        />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  test("should call MovementsRepository.query with limit and order", async () => {
    const component = render(
      <MovementsMostRecents
        movementsRepository={useNodeMovementsRepository.useNodeMovementsRepository()}
      />
    );

    await waitFor(async () => {
      expect((await component.findAllByRole("listitem")).length).toBe(
        movementList.length
      );
    });

    expect(movementsRepository.query).toHaveBeenCalledWith({
      limit: 10,
      order: OrderType.DESC,
    });
    expect(movementsRepository.query).toHaveBeenCalledTimes(1);
    expect(movementListStore.updateMovementList).toHaveBeenCalledTimes(1);
    expect(movementListStore.updateMovementList).toHaveBeenCalledWith(
      movementList
    );
  });
});
