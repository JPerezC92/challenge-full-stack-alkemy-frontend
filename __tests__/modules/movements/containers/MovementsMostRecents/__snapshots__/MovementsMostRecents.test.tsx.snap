// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovementsMostRecents container should match the snapshot 1`] = `
<section
  className="rounded-md bg-gradient-to-b from-indigo-100 to-indigo-50 p-4"
>
  <h2
    className="mb-4 text-lg font-bold"
  >
    Recent movements
  </h2>
  <ol
    className="grid grid-rows-[10] gap-4 sm:grid-cols-2 sm:grid-rows-5"
  >
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-1
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            1000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-2
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            2000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-3
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            3000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-4
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            4000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-5
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            5000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-6
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            6000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-7
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            7000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-8
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            8000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-9
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            9000.00
          </b>
        </span>
      </div>
    </li>
    <li
      className="flex min-w-max flex-col gap-1 rounded-sm border border-cyan-500/30 p-3 undefined"
    >
      <header>
        <span
          className="text-lg font-bold capitalize"
        >
          move
        </span>
      </header>
      <div
        className=" flex items-center justify-between gap-4 text-right text-lg"
      >
        <span
          className="min-w-max text-sm font-semibold italic text-gray-400"
        >
          2020-01-10
        </span>
        <span
          className="text-orange-600 min-w-max"
        >
          <b>
            -
             $
            10000.00
          </b>
        </span>
      </div>
    </li>
  </ol>
</section>
`;
